source/crud/crud.rst

```diff
@@ -137,6 +137,8 @@ Read
      * MAY setup a cursor to be executed upon iteration against the $out collection such
      * that if a user were to iterate a pipeline including $out, results would be returned.
      *
+     * Note: implementations should return a cursor for Iterable<Document>.
```

Comment:
If we're talking about something like Java's Iterable, which is defined as

    public interface Iterable<T> {
        Iterator<T> iterator();
    }

then an Iterable can be iterated multiple times, so it can't also be a cursor.  Rather, it's an object that can produce an Iterator, and it's each Iterator that is actually backed by a cursor.  Java and .NET drivers override this iterator() method to return, effectively, a CloseableIterator, so that applications can ensure that the cursor is closed even when the Iterator is not exhausted.

We have two options here:

1. Change the note to say something like "Iterable<Document> instances returned by this method should produce iterators that are backed by a cursor"
2. Change the return type to Iterator<Document>.  Whether a driver produces Iterables that in turn produce Iterators can be treated as an acceptable deviation based on language idioms.

In both cases the Iterator must have some way to be closed (e.g. in Java by implementing the AutoCloseable interface)

I also think that distinct should say the exact same thing.  The fact that distinct currently does not return a cursor is an implementation detail that may change.  We got burned by this a while ago when aggregate started returning a cursor, and I can see distinct going the same way, say by being re-implemented as an aggregation pipeline.  But even if that doesn't happen there is no reason an application need be aware of the difference.

Reply:
I prefer the language stating that the iterator should be backed by a cursor.

I think it's more flexible to avoid referring to Iterator, since that's going to be a type in most languages and the CRUD spec already defines "Iterable" in its terms for this purpose.

I would change this to:

```
Note: the Iterable<Document> return value should be backed by a cursor.
```